src/engagement_db_to_analysis/rapid_pro_advert_functions.py

```diff
@@ -0,0 +1,271 @@
+    :param type: AnalysisCache
+    :param target_uuids: Set containing all uuids for the target context e.g opt_out uuids.
+    :type target_uuids: set of str
+    :param group_name: Name of the group to sync in rapid_pro.
+    :type group_name: str
+    :param uuid_table: UUID table to use to de-identify contact urns.
+    :type uuid_table: id_infrastructure.firestore_uuid_table.FirestoreUuidTable.
+    :param rapid_pro: Rapid Pro client to sync the groups to.
+    :type rapid_pro: rapid_pro_tools.rapid_pro.RapidProClient
+    '''
+
+    synced_uuids = []
+    if cache is not None:
+        synced_dataset_nc_uuids = cache.get_synced_uuids(group_name)
+        log.info(f'Found {len(synced_dataset_nc_uuids)} previously uploaded to {group_name}...')
+
+    # If cache is available, check for uuids to sync in the current pipeline run.
+    uuids_to_sync = _get_uuids_to_sync(target_uuids, synced_uuids)
+
+    # Re-identify the uuids.
+    urns_to_sync = _convert_uuids_to_urns(uuids_to_sync, uuid_table)
+
+    # Sync the group contacts to rapid_pro.
+    log.info(f'Adding {len(urns_to_sync)} contacts to {group_name} group...')
+    advert_group_uuid = _ensure_rapid_pro_group_exists(group_name, rapid_pro)
+
+    for urn in urns_to_sync:
+        _update_group_for_urn(urn, advert_group_uuid, rapid_pro)
+        synced_uuids.append(uuid_table.data_to_uuid(urn))
+
+    if cache is not None:
+        cache.set_synced_uuids(group_name, synced_uuids)
+
+def sync_advert_contacts_to_rapidpro(participants_by_column, uuid_table, pipeline_config, rapid_pro,
+                         google_cloud_credentials_file_path, membership_group_dir_path, cache_path):
+
+    if cache_path is None:
+        cache = None
+        log.warning(f"No `cache_path` provided. This tool will perform a full sync of advert contacts to rapidpro")
+    else:
+        log.info(f"Initialising EngagementAnalysisCache at '{cache_path}'")
+        cache = AnalysisCache(f"{cache_path}")
+
+    opt_out_uuids, weekly_advert_uuids = _generate_weekly_advert_and_opt_out_uuids(participants_by_column,
+                                                                                pipeline_config.analysis,
+                         google_cloud_credentials_file_path, membership_group_dir_path)
+
+    log.info(f'Syncing consent_withdrawn contact_fields in rapidpro... ')
+    #check for opt out uuids to sync in this pipeline run
+    synced_opt_out_uuids = []
+    if cache is not None:
+        synced_opt_out_uuids = cache.get_synced_uuids('opt_out_uuids')
+        log.info(f"Found {len(synced_opt_out_uuids)} previously synced consent withdrawn uuids...")
+        
+    opt_out_uuids_to_sync = _get_uuids_to_sync(opt_out_uuids, synced_opt_out_uuids)
+
+    # Re-identify the uuids.
+    opt_out_urns = _convert_uuids_to_urns(opt_out_uuids_to_sync, uuid_table)
+
+    # Update consent_withdrawn contact field for opt_out contacts
+    log.info(f'Found {len(opt_out_urns)} new opt_out contacts to sync ')
```

Comment:
Why are we syncing consent_withdrawn when it's also being done in the engagement db -> rapid pro sync?

Reply:
So that we can always be sure we have the updated consent_withdrawn status for advert flows even for the cases where a pipeline engagement DB -> rapid pro sync is not always set to True e.g when this is done periodically after the end of a season.